src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+{
+    module_state *state = (module_state *)userdata;
+    // a lot of the code below has been taken from the playing a tone pipewire tutorial, see the pipewire documentation for more details
+
+    struct pw_buffer *b;
+    struct spa_buffer *buf;
+    uint32_t *destination_memory, read_index, available_for_loading, must_load_with_silence;
+    int32_t available_samples, number_of_samples;
+    if ((b = pw_stream_dequeue_buffer(state->stream)) == NULL)
+    {
+        pw_log_warn("out of buffers: %m");
+        return;
+    }
+    buf = b->buffer;
+    // pipewire's buffers are mapped directly from the server to the client
+    // As such, if the memory we're given doesn't exist, we have to abort imediately, because doing otherwise most certainly would lead to undefined behaviour
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    number_of_samples = buf->datas[0].maxsize; // how many samples, in bytes,  of audio are in this chunk. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases, for example buggy hardware drivers which don't probe the card correctly, so on_process gets called very often, with a value of requested samples greatter than what pipewire could allocate, in maxsize
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        number_of_samples = SPA_MIN(number_of_samples, b->requested * state->stride); // because this value is given in frames, and we want samples
+    }
+#endif
+
+    // if there's something to be read from the ring buffer at this time, we do so now
+    if (available_samples > 0)
+    {
+        available_for_loading = SPA_MIN(available_samples, number_of_samples);
+    }
+    else
+    {
+        available_for_loading = 0;
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = number_of_samples - available_for_loading;
+    if (available_for_loading > 0)
+    {
+        spa_ringbuffer_read_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, read_index, destination_memory, available_for_loading);
```

Comment:
`read_index` needs a `%SAMPLE_BUFFER_SIZE`, otherwise you will be reading garbage